app/models/service_template.rb

```diff
@@ -103,12 +88,31 @@ def self.with_additional_tenants
     references(table_name, :tenants).includes(:service_template_tenants => :tenant)
   end
 
+  def self.all_catalog_item_types
+    @all_catalog_item_types ||= begin
+      builtin_catalog_item_types = {
+        "generic"               => N_("Generic"),
+        "generic_orchestration" => N_("Orchestration"),
+      }
+
+      ExtManagementSystem.supported_types_for_catalog
+        .flat_map(&:catalog_types)
+        .reduce(builtin_catalog_item_types, :merge)
+    end
+  end
+
   def self.catalog_item_types
-    ci_types = Set.new(Rbac.filtered(ExtManagementSystem.all).flat_map(&:supported_catalog_types))
+    ci_types = Set.new(
+      Rbac.filtered(ExtManagementSystem.all)
+        .select(&:supported_for_catalog?)
+        .flat_map(&:catalog_types).reduce({}, :merge).keys
```

Comment:
I take it we are assuming a low number of Ems per installation. (seams reasonable)

Reply:
Hm so you did get me thinking, this was always a class attribute (it was an instance method before but it was hard coded per-provider-plugin).

We don't need to go through all the EMSs we just need to get the unique set of EMS classes